server/src/main/java/com/linecorp/centraldogma/server/internal/api/CredentialServiceV1.java

```diff
@@ -0,0 +1,116 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.centraldogma.server.internal.api;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import com.linecorp.armeria.server.annotation.ConsumesJson;
+import com.linecorp.armeria.server.annotation.ExceptionHandler;
+import com.linecorp.armeria.server.annotation.Get;
+import com.linecorp.armeria.server.annotation.Param;
+import com.linecorp.armeria.server.annotation.Post;
+import com.linecorp.armeria.server.annotation.ProducesJson;
+import com.linecorp.armeria.server.annotation.Put;
+import com.linecorp.armeria.server.annotation.StatusCode;
+import com.linecorp.centraldogma.common.Author;
+import com.linecorp.centraldogma.internal.api.v1.PushResultDto;
+import com.linecorp.centraldogma.server.command.CommandExecutor;
+import com.linecorp.centraldogma.server.internal.api.auth.RequiresReadPermission;
+import com.linecorp.centraldogma.server.internal.api.auth.RequiresWritePermission;
+import com.linecorp.centraldogma.server.internal.storage.project.ProjectApiManager;
+import com.linecorp.centraldogma.server.mirror.MirrorCredential;
+import com.linecorp.centraldogma.server.storage.project.Project;
+import com.linecorp.centraldogma.server.storage.repository.MetaRepository;
+
+/**
+ * Annotated service object for managing credential service.
+ */
+@ProducesJson
+@ExceptionHandler(HttpApiExceptionHandler.class)
+public class CredentialServiceV1 extends AbstractService {
+
+    private final ProjectApiManager projectApiManager;
+
+    public CredentialServiceV1(ProjectApiManager projectApiManager, CommandExecutor executor) {
+        super(executor);
+        this.projectApiManager = projectApiManager;
+    }
+
+    /**
+     * GET /projects/{projectName}/credentials
+     *
+     * <p>Returns the list of the credentials in the project.
+     */
+    @RequiresReadPermission(repository = Project.REPO_META)
+    @Get("/projects/{projectName}/credentials")
+    public CompletableFuture<List<MirrorCredential>> listCredentials(@Param String projectName) {
+        return metaRepo(projectName).credentials();
+    }
+
+    /**
+     * GET /projects/{projectName}/credentials/{id}
+     *
+     * <p>Returns the credential for the ID in the project.
+     */
+    @RequiresReadPermission(repository = Project.REPO_META)
+    @Get("/projects/{projectName}/credentials/{id}")
+    public CompletableFuture<MirrorCredential> getCredentialById(@Param String projectName, @Param String id) {
+        return metaRepo(projectName).credential(id);
+    }
+
+    /**
+     * POST /projects/{projectName}/credentials
+     *
+     * <p>Creates a new credential.
+     */
+    @RequiresWritePermission(repository = Project.REPO_META)
+    @Post("/projects/{projectName}/credentials")
+    @ConsumesJson
+    @StatusCode(201)
+    public CompletableFuture<PushResultDto> createCredential(@Param String projectName,
```

Comment:
So users always have to create the credential first before creating a mirror configuration.
Would it be worth if we provide an API that commits both together at once ?

Reply:
A credential can be used in multiple mirror configs so the relation is 1(credential):N(mirror).